server/src/main/java/io/crate/execution/engine/aggregation/impl/AverageAggregation.java

```diff
@@ -100,6 +104,21 @@ public Double value() {
             }
         }
 
+        public void addNumber(double number, boolean isIntegral) {
+            this.sum = isIntegral ? this.sum + number : kahanSummationForDouble.sum(this.sum, number);
+            this.count++;
+        }
```

Comment:
Sorry I was probably too vague in my last comment. I meant that it could parameterize the functions itself. 

Not sure if it has any performance impact, but it could avoid the branching.

Reply:
I did another implementation (third commit), here is benchmark results:

current implementation (with branching)

```
  Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:08<00:00, 58.03 requests/s]
Runtime (in ms):
    mean:    16.172 ± 2.325
    min/max: 13.883 → 606.535
Percentile:
    50:   14.773 ± 26.522 (stdev)
    95:   16.126
    99.9: 606.535
```
    
  Third commit (without branching)
```
  Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:09<00:00, 50.30 requests/s]
Runtime (in ms):
    mean:    18.553 ± 2.414
    min/max: 15.098 → 628.538
Percentile:
    50:   16.483 ± 27.535 (stdev)
    95:   21.320
    99.9: 628.538
```

It got worse, probably because of extra unboxing in Consumers.

I didn't do smth like `addNumber(double num)` and `addNumber(long num)` since this PR keeps double sum for the sake of BWC (overflow addressed in another PR with numeric type) and I anyway have to do `Number.doubleValue`. another reason for having consumers rather than 2 state changers is that reduce accepts AverageState and I have to do branching there anyway.

Reply:
Why don't you use an interface that doesn't require boxing?

Reply:
I tried FunctionalINterface with primitive argument instead of Consumer (if I got suggestion correctly) but it got even worth

```
 Statement:
     select avg("adRevenue") from uservisits
   Concurrency: 1
   Iterations: 500
100%| 500/500 [00:12<00:00, 40.21 requests/s]
Runtime (in ms):
    mean:    23.574 ± 2.280
    min/max: 19.538 → 597.614
Percentile:
    50:   21.541 ± 26.012 (stdev)
    95:   26.650
    99.9: 597.614
```
    
It looks like some predefined caller (Consumer or Interface) --> call relevant AverageState method is slower than if + direct relevant method call

Reply:
Thanks for investigating. Let's go with the initial variant then.

Reply:
👌